v1/status_message.go

```diff
@@ -0,0 +1,142 @@
+package statusproto
+
+import (
+	"crypto/ecdsa"
+	"github.com/pkg/errors"
+	"log"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/golang/protobuf/proto"
+	"github.com/jinzhu/copier"
+	"github.com/status-im/status-protocol-go/applicationmetadata"
+	"github.com/status-im/status-protocol-go/datasync"
+	"github.com/status-im/status-protocol-go/encryption"
+	whisper "github.com/status-im/whisper/whisperv6"
+)
+
+// StatusMessage is any Status Protocol message.
+type StatusMessage struct {
+	// TransportMessage is the parsed message received from the trasport layer, i.e the input
+	TransportMessage *whisper.Message
+	// ParsedMessage is the parsed message by the application layer, i.e the output
+	ParsedMessage interface{}
+
+	// TransportPayload is the payload as received from the transport layer
+	TransportPayload []byte
+	// DecryptedPayload is the payload after having been processed by the encryption layer
+	DecryptedPayload []byte
+
+	// ID is the canonical ID of the message
+	ID []byte
+	// Hash is the transport layer hash
+	Hash []byte
+
+	// TransportLayerSigPubKey contains the public key provided by the transport layer
+	TransportLayerSigPubKey *ecdsa.PublicKey
+	// ApplicationMetadataLayerPubKey contains the public key provided by the application metadata layer
+	ApplicationMetadataLayerSigPubKey *ecdsa.PublicKey
+}
+
+// SigPubKey returns the most important signature, from the application layer to transport
+func (s *StatusMessage) SigPubKey() *ecdsa.PublicKey {
+	if s.ApplicationMetadataLayerSigPubKey != nil {
+		return s.ApplicationMetadataLayerSigPubKey
+	}
+
+	return s.TransportLayerSigPubKey
+}
+
+func (s *StatusMessage) Clone() (*StatusMessage, error) {
+	copy := &StatusMessage{}
+
+	err := copier.Copy(&copy, s)
+	return copy, err
+}
+
+func (m *StatusMessage) HandleTransport(shhMessage *whisper.Message) error {
```

Comment:
I wonder if making `m.Handle*` methods private and moving `StatusMessage` to the root of the project wouldn't make more sense. `StatusMessage` is not really something described in the spec but an implementation detail.

Reply:
I might do this in a separate PR, once we refactor the sending flow as well, so we have a bit more context